src/utils/atomWithObservable.ts

```diff
@@ -56,113 +58,98 @@ export function atomWithObservable<TData>(
-
-    let emittedValueBeforeMount: TData | Promise<TData> | typeof EMPTY = EMPTY
-    let isSync = true
-    let setData: (data: TData | Promise<TData>) => void = (data) => {
-      // First we set the initial value (if not other initialValue was provided)
-      // All the following data is saved in a variable so it doesn't get lost before the mount
-      if (options?.initialValue === undefined && !initialValueWasEmitted) {
-        if (isSync) {
-          initialEmittedValue = data
-        }
-        resolveEmittedInitialValue?.(data)
-        initialValueWasEmitted = true
-        resolveEmittedInitialValue = null
-      } else {
-        emittedValueBeforeMount = data
-      }
-    }
-
-    const dataListener = (data: TData) => {
-      setData(data)
+    type Result = { d: Data } | { e: AnyError }
+    let resolve: ((result: Result) => void) | undefined
+    const makePending = () =>
+      new Promise<Result>((r) => {
+        resolve = r
+      })
+    const initialResult: Result | Promise<Result> =
+      options && 'initialValue' in options
+        ? {
+            d:
+              typeof options.initialValue === 'function'
+                ? (options.initialValue as () => Data)()
+                : (options.initialValue as Data),
+          }
+        : makePending()
+
+    let setResult: ((result: Result) => void) | undefined
+    let lastResult: Result | undefined
+    const listener = (result: Result) => {
+      lastResult = result
+      resolve?.(result)
+      setResult?.(result)
     }
 
-    const errorListener = (error: unknown) => {
-      setData(Promise.reject<TData>(error))
-    }
-
-    let subscription: Subscription | null = null
-    let initialValue:
-      | TData
-      | Promise<TData>
-      | InitialValueFunction<TData>
-      | undefined
-    if (options?.initialValue !== undefined) {
-      initialValue = getInitialValue(options)
-    } else {
-      // FIXME
-      // There is the potential for memory leaks in this implementation.
-      //
-      // If the observable doesn't emit an initial value before the component that uses the atom gets destroyed,
-      // the onMount function never gets called and therefore the subscription never gets cleaned up.
-      //
-      // Unfortunately, currently there is no good way to prevent this issue (as of 2022-05-23).
-      // Timeouts may lead to an endless loading state, if the subscription get's cleaned up too quickly.
-      //
-      // Discussion: https://github.com/pmndrs/jotai/pull/1170
-      subscription = observable.subscribe(dataListener, errorListener)
-      initialValue = initialEmittedValue
+    let subscription: Subscription | undefined
+    let timer: Timeout | undefined
+    const isNotMounted = () => !setResult
+    const start = () => {
+      if (subscription) {
+        clearTimeout(timer)
+        subscription.unsubscribe()
+      }
+      subscription = observable.subscribe(
+        (d) => listener({ d }),
+        (e) => listener({ e })
```

Comment:
Seems like this style is deprecated in rxjs. https://rxjs.dev/deprecations/subscribe-arguments